syntax = "proto3";

package kart.sensors;

// Sensor command IDs (sensor types)
// These are used as command_id in the protocol
enum SensorCommandId {
  TEMPERATURE = 0;          // Temperature sensors
  RPM = 1;                  // RPM sensors
  VOLTAGE = 2;              // Voltage sensors
  CURRENT = 3;              // Current sensors
  PRESSURE = 4;             // Pressure sensors
  POSITION = 5;             // Position sensors
  ACCELERATION = 6;         // Accelerometer
  GYROSCOPE = 7;            // Gyroscope
  SAFETY = 8;               // Safety-critical sensors
  STATUS = 9;               // Sensor status message
}

// Sensor status values
// Used in value field when command_id is STATUS
enum SensorStatusValue {
  OK = 0;                   // Normal operation
  ERROR_CONNECTION = 1;     // Connection error
  ERROR_TIMEOUT = 2;        // Timeout error
  ERROR_RANGE = 3;          // Out of range error
  ERROR_CALIBRATION = 4;    // Calibration error
  DISABLED = 5;             // Sensor is disabled
  INITIALIZING = 6;         // Sensor is initializing
  ERROR_GENERAL = 7;        // General error
}

// Temperature sensor location
// Used in component_id field when command_id is TEMPERATURE
enum TemperatureSensorLocation {
  MOTOR = 0;                // Motor temperature
  CONTROLLER = 1;           // Controller temperature
  BATTERY = 2;              // Battery temperature
  AMBIENT = 3;              // Ambient temperature
  COOLANT = 4;              // Coolant temperature
}

// RPM sensor location
// Used in component_id field when command_id is RPM
enum RpmSensorLocation {
  WHEEL_FRONT_LEFT = 0;     // Front left wheel
  WHEEL_FRONT_RIGHT = 1;    // Front right wheel
  WHEEL_REAR_LEFT = 2;      // Rear left wheel
  WHEEL_REAR_RIGHT = 3;     // Rear right wheel
  ENGINE = 4;               // Engine RPM
  MOTOR_MAIN = 5;           // Main drive motor
  MOTOR_AUXILIARY = 6;      // Auxiliary motor
}
